### ISSUE TITLE
Hyper-V isolation with bind mount appears to hold unnecessary file locks on container host

### ISSUE THREAD

**Issue by nickwesselman** [moby/moby#42803]
Moving to moby at the request of @stephen-turner.

  - [X] I have tried with the latest version of Docker Desktop
  - [X] I have tried disabling enabled experimental features
  - [X] I have uploaded Diagnostics
  - Diagnostics ID: 953333DF-EA30-47EC-A977-EEAE6AE356DE/20210830154356

### Actual behavior
I'm seeing behavior myself and [reported by others](https://stackoverflow.com/questions/66752750/visual-studio-build-fails-while-copying-files-to-the-bin-directory-due-to-file-l) that a bind mount in a Windows container w/ Hyper-V isolation can result in files seemingly being locked by the Hyper-V VM, even though there is no process on the container which is locking them. This is definitely a change in a recent Docker (or Windows?) version update, as this worked just fine previously.

In my case running `dotnet watch` in the container, the initial build of the mounted source works. After a file change, `dotnet watch` will kill the running app, and attempt to rebuild. The subsequent builds all fail however:

> CSC : error CS2012: Cannot open 'C:\src\obj\container\Debug\net5.0\dotnet-docker-lock-issue.dll' for writing -- 'The process cannot access the file 'C:\src\obj\container\Debug\net5.0\dotnet-docker-lock-issue.dll' because it is being used by another process.' [C:\src\dotnet-docker-lock-issue.csproj]

Using sysinternals `handle`, I see no processes on the container which have locks on this path.

```
PS C:\> handle64 -accepteula C:\src\obj\

Nthandle v4.22 - Handle viewer
Copyright (C) 1997-2019 Mark Russinovich
Sysinternals - www.sysinternals.com

No matching handles found.
```

However there is a lock on the container host:

```
PS C:\> handle64.exe C:\dev\dotnet-docker-lock-issue\obj\

Nthandle v4.22 - Handle viewer
Copyright (C) 1997-2019 Mark Russinovich
Sysinternals - www.sysinternals.com

vmwp.exe           pid: 17640  type: File          35DC: C:\dev\dotnet-docker-lock-issue\obj\container\Debug\net5.0\d377a3cb-bad4-4c2d-8e66-e6a46db6ba2c_dotnet-docker-lock-issue.dll
```

Why does the Hyper-V process have this lock?

I've created a simple example which reproduces the behavior [in this repository](https://github.com/nickwesselman/dotnet-docker-lock-issue).

### Expected behavior
Killed processes in a Windows container don't continue to hold file locks.

### Information

  - Windows Version: 21H1 (OS Build 19043.985)
  - Docker Desktop Version: 3.60 (67351)
  - WSL2 or Hyper-V backend?  Hyper-V
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

### Steps to reproduce the behavior
Documented [in this repository](https://github.com/nickwesselman/dotnet-docker-lock-issue)

**Comment by nickwesselman** [comment]
Moved from docker/for-win#12028

**Comment by nickwesselman** [comment]
Added `docker info` and `docker version`

**Comment by nickwesselman** [comment]
Would HCS traces be useful on this one? It's pretty easy to reproduce.

**Comment by nickwesselman** [comment]
@CaffeineBytes No, have not had time to investigate further and never got a response on whether HCS would be helpful here

**Comment by katiewasnothere** [comment]
@nickwesselman Yes traces would be great

**Comment by nickwesselman** [comment]
Hi @katiewasnothere! I just tried to use HcsTrace, but I received an error when stopping.

```
➜ wpr.exe -start HcsTraceProfile.wprp!HcsArgon -filemode
➜ wpr.exe -stop HcsTrace.etl "Hyper-V File Lock Issue"

        Cannot change thread mode after it is set.

        Profile Id: RunningProfile

        Error code: 0x80010106
```

Any ideas?

**Comment by katiewasnothere** [comment]
@nickwesselman hmm, I've never seen that. The only thing I could find was this https://devblogs.microsoft.com/performance-diagnostics/wpr-start-and-stop-commands/. Does that resolve the issue?

**Comment by CaffeineBytes** [comment]
I wanted to check to see if those files ended up helping determine what is causing the lock. Is there any more information that would help, or something I can try?

**Comment by katiewasnothere** [comment]
@nickwesselman Do you know what OS build worked previously?

**Comment by nickwesselman** [comment]
@katiewasnothere I believe I was on 1909 previously. Not sure if it was a Docker or Windows upgrade that caused the issue.

**Comment by pbering** [comment]
Any update on this issue? Can we do anything to help?

**Comment by jakoblover** [comment]
Having the same issue on my end

**Comment by travnick** [comment]
I have similar issues with vmwp.exe locking the files. I see different random errors on CI like `Failed to delete empty directory:`, `Failed to delete file:`, `The process cannot access the file because it is being used by another process.`, `Access to the path 'path here' is denied.`.
While try to manually remove content of mounted volume, I'm facing an error:
![image](https://user-images.githubusercontent.com/619590/225868908-d575fb1b-f602-484d-aff1-f13b6611f2ad.png)

But in my case `handle.exe` (Sysinternals) shows nothing (nothing about this particular file/directory, both on host and inside the container).

```
docker version
Client:
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:50:24 2023
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.24)
  Go version:       go1.19.5
  Git commit:       bc3805a0
  Built:            Thu Feb  9 19:46:43 2023
  OS/Arch:          windows/amd64
  Experimental:     false
  ```

```
docker info
Client:
 Context:    default
 Debug Mode: false

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 23.0.1
 Storage Driver: windowsfilter
  Windows:
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics internal l2bridge l2tunnel nat null overlay private transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
 Swarm: inactive
 Default Isolation: hyperv
 Kernel Version: 10.0 19044 (19041.1.amd64fre.vb_release.191206-1406)
 Operating System: Microsoft Windows Version 21H2 (OS Build 19044.2728)
 OSType: windows
 Architecture: x86_64
 CPUs: 8
 Total Memory: 31.68GiB
 Name: MY_PC_NAME
 ID: JFPK:B4YQ:VW2T:E2R5:2S46:JDUT:MKDI:XM5Y:ZOH2:PPT5:6VET:HTLY
 Docker Root Dir: C:\ProgramData\docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
 ```

**Comment by x3nb63** [comment]
here the `vmvp.exe` appears to just pick the .exe files for locking, while all other files on the "mount" are unaffected.

can some content scanning layer be involved here? This Windows Defender construction maybe?  (i have no third party virus software installed)

I don't know where to look / how to find out.

**Comment by dzanitchkhahMC** [comment]
Hi,

I may be barking up the wrong tree here but...

I believe we're experiencing this issue after upgrading from 10.0 to 10.3. My initial assumption was that the problem was related to the upgrade (which included upgrading the rendering host). 

I don't believe our docker versions changed during the upgrade period and it affected all the devs... The only thing that changed during this time (other than the rendering host version) was this in the .env file...

```
SITECORE_VERSION=10.0.0-20H2
SOLUTION_BUILD_IMAGE=mcr.microsoft.com/dotnet/framework/sdk:4.8-20210713-windowsservercore-20H2
SOLUTION_BASE_IMAGE=mcr.microsoft.com/windows/nanoserver:20H2
NETCORE_BUILD_IMAGE=mcr.microsoft.com/dotnet/sdk:3.1.411-nanoserver-20H2

SITECORE_VERSION=10.3.1-ltsc2019
SOLUTION_BUILD_IMAGE=mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2019
SOLUTION_BASE_IMAGE=mcr.microsoft.com/windows/nanoserver:1809
NETCORE_BUILD_IMAGE=mcr.microsoft.com/dotnet/sdk:6.0-windowsservercore-ltsc2019
```

If this isn't a Sitecore problem, that would suggest it's a windows container problem... Is there any way would change to a different Windows build? From memory I think Sitecore might have pulled some windows builds?

**Comment by Sam13** [comment]
I also experience this problem a lot when doing continuous integration builds with Windows containers.

As a workaround I copy the contents of the bind mount into a directory in the container before executing my Docker command.
After command execution (which ran on the directory in the container) I (partially) copy the contents back to the bind mount.

This may work for certain scenarios.

**Comment by andrea-masciotta-idsg** [comment]
Any update on this? I also encountered this problem when trying to run a CI build on Windows, using Docker and Jenkins. The build tries to overwrite some file in the build folder (which contains my code and is mounted by Jenkins from the host) but fails because access is denied.

When trying to reproduce the issue locally, stopping and restarting the container fixes the problem, but of course this is not an option during a CI build.
